ToolBox/BovespaDownloader/BovespaDataDownloader.cs

```diff
@@ -0,0 +1,456 @@
+                    zip.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Aggregates a list of ticks at the requested resolution
+        /// </summary>
+        /// <param name="symbol">Symbol</param>
+        /// <param name="ticks">Input tick data</param>
+        /// <param name="resolution">Output resolution</param>
+        /// <returns>Enumerable of trade bars or quote bars</returns>
+        private static IEnumerable<BaseData> AggregateTicks(Symbol symbol, IEnumerable<BaseData> data, Resolution resolution)
+        {
+            var ticks = data.Cast<Tick>().OrderBy(t => t.Time).ToList();
+
+            if (ticks.All(t => t.TickType == TickType.Trade))
+            {
+                return
+                    (from t in ticks
+                     group t by t.Time.RoundDown(resolution.ToTimeSpan())
+                         into g
+                         select new TradeBar
+                         {
+                             DataType = MarketDataType.TradeBar,
+                             Symbol = symbol,
+                             Time = g.Key,
+                             Open = g.First().LastPrice,
+                             High = g.Max(t => t.LastPrice),
+                             Low = g.Min(t => t.LastPrice),
+                             Close = g.Last().LastPrice,
+                             Volume = g.Sum(t => t.Quantity),
+                         });
+            }
+            else
+            {
+                return 
+                    (from t in ticks
+                     group t by t.Time.RoundDown(resolution.ToTimeSpan()) into g
+                     
+                     select new QuoteBar
+                     {
+                         DataType = MarketDataType.QuoteBar,
+                         Symbol = symbol,
+                         //Time = g.Key,
+                         //AvgBidSize = (long)g.Where(t => t.BidSize > 0).Average(t => t.BidSize),
+                         //AvgAskSize = (long)g.Where(t => t.AskSize > 0).Average(t => t.AskSize),
+                         //Bid = new Bar
+                         //(
+                         //    g.Where(t => t.BidSize > 0).First().LastPrice,
+                         //    g.Where(t => t.BidSize > 0).Max(t => t.LastPrice),
+                         //    g.Where(t => t.BidSize > 0).Min(t => t.LastPrice),
+                         //    g.Where(t => t.BidSize > 0).Last().LastPrice
+                         //),
+                         //Ask = new Bar
+                         //(
+                         //    g.Where(t => t.AskSize > 0).First().LastPrice,
+                         //    g.Where(t => t.AskSize > 0).Max(t => t.LastPrice),
+                         //    g.Where(t => t.AskSize > 0).Min(t => t.LastPrice),
+                         //    g.Where(t => t.AskSize > 0).Last().LastPrice
+                         //),         
+                     });
```

Comment:
Hey @mchandschuh, I don't know how to do this properly (so I commented out).
I have ticks with bid data (`<BidSize > 0 && AskSize = 0`>) and ask data (`<BidSize = 0 && AskSize > 0`>). I want to group then together by `<Time.RoundDown(...)`> then I need to group them by bid-only and ask-only list to make the calculations.
I don't think my solution is right, because the `<Where(...)`> can return a zero-length list and that would result into an exception when I try to access `<First()`> and`<Last()`>. And, of course, the several `<Where(...)`> calls are redundant and probably inefficient.